pkg/i2gw/ingress2gateway.go

```diff
@@ -35,70 +35,92 @@ import (
 	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
 )
 
-func Run(printer printers.ResourcePrinter, namespace string, inputFile string) {
+func Run(printer printers.ResourcePrinter, namespace string, inputFile string) error {
 	conf, err := config.GetConfig()
 	if err != nil {
-		fmt.Println("failed to get client config")
-		os.Exit(1)
+		return fmt.Errorf("failed to get client config: %w", err)
 	}
 
 	cl, err := client.New(conf, client.Options{})
 	if err != nil {
-		fmt.Println("failed to create client")
-		os.Exit(1)
+		return fmt.Errorf("failed to create client: %w", err)
 	}
 	cl = client.NewNamespacedClient(cl, namespace)
 
-	ingressList := &networkingv1.IngressList{}
-
+	var ingresses networkingv1.IngressList
 	if inputFile != "" {
-		err = constructIngressesFromFile(ingressList, inputFile)
-		if err != nil {
-			fmt.Printf("failed to open input file: %v\n", err)
-			os.Exit(1)
+		if err = constructIngressesFromFile(&ingresses, inputFile); err != nil {
+			return fmt.Errorf("failed to open input file: %w", err)
 		}
 	} else {
-		err = cl.List(context.Background(), ingressList)
-		if err != nil {
-			fmt.Printf("failed to list ingresses: %v\n", err)
-			os.Exit(1)
+		if err = cl.List(context.Background(), &ingresses); err != nil {
+			return fmt.Errorf("failed to list ingresses: %w", err)
 		}
 	}
 
-	if len(ingressList.Items) == 0 {
+	if len(ingresses.Items) == 0 {
 		msg := "No resources found"
 		if namespace != "" {
 			fmt.Printf("%s in %s namespace\n", msg, namespace)
 		} else {
 			fmt.Println(msg)
 		}
-		return
+		return nil
 	}
 
-	httpRoutes, gateways, errors := ingresses2GatewaysAndHTTPRoutes(ingressList.Items)
-	if len(errors) > 0 {
-		fmt.Printf("# Encountered %d errors\n", len(errors))
-		for _, err = range errors {
+	var gateways []gatewayv1beta1.Gateway
+	var httpRoutes []gatewayv1beta1.HTTPRoute
+	var errs field.ErrorList
+
+	providerByName := constructProviders(&ProviderConf{})
+	if len(providerByName) == 0 {
+		return fmt.Errorf("no providers")
+	}
+
+	resources := IngressResources{Ingresses: ingresses.Items}
+
+	for name, provider := range providerByName {
+		if err = provider.ReadResourcesFromCluster(context.Background(), &resources.CustomResources); err != nil {
```

Comment:
maybe init context outside and pass the same one here?

Reply:
done